### PR TITLE
Fix #50: Check pull request for job files

### LINKED ISSUE #50: Oppiabot should detect PRs which affect datastore layers
Add automation to detect any PRs which affect the datastore layer (maybe using github actions or oppiabot).

### PR DESCRIPTION
**Pull request by jameesjohn**
## Explanation
This PR checks for modification of job files and the job registry in a pull request.

This PR has been tested at https://github.com/jameesjohn/comment-on-pr/pull/31, https://github.com/jameesjohn/comment-on-pr/pull/32

## Checklist
- [x] I have successfully deployed my own instance of Oppiabot.
  - You can find instructions for doing this [here](https://github.com/oppia/oppiabot/wiki/Deploying-your-own-instance-of-the-oppiabot).
- [x] I have manually tested all the changes made in this PR following the [manual tests matrix](https://github.com/oppia/oppiabot/wiki/Manual-Tests-Matrix).

### CODE REVIEW

File: lib/checkPullRequestJob.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
+const {SERVER_JOBS_ADMIN} = require('../userWhitelist.json');
+/**
+ * @param {string} filename
+ */
+const isJobFile = (filename) => {
+  const REGISTRY_FILENAME = 'jobs_registry.py';
+  const JOB_FILE_FIX = '_jobs_';
```
Comment (ankita240796) [inline lib/checkPullRequestJob.js:7]: Define these outside the function.

Comment (ankita240796) [inline lib/checkPullRequestJob.js:7]: What does FIX represent here?

File: lib/checkPullRequestJob.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  });
+
+  if (containsJobFile) {
+    const prAuthor = pullRequest.user.login;
+    const serverJobsForm = 'server jobs form'.link(
+      'https://goo.gl/forms/XIj00RJ2h5L55XzU2');
+    const commentBody = 'Hi @' + SERVER_JOBS_ADMIN + ', PTAL at this PR, ' +
+      'it adds/modifies a server job. Also @' + prAuthor + ', please ' +
+      'endeavour to fill the ' + serverJobsForm + ' for the new job ' +
+      'to get tested on the test server. Thanks!';
```
Comment (ankita240796) [inline lib/checkPullRequestJob.js:46]: to get tested on the test server -> to be tested on the backup server. This PR can be merged only after the test is successful.

File: lib/checkPullRequestJob.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    return isJobFile(file.filename);
+  });
+
+  if (containsJobFile) {
+    const prAuthor = pullRequest.user.login;
+    const serverJobsForm = 'server jobs form'.link(
+      'https://goo.gl/forms/XIj00RJ2h5L55XzU2');
+    const commentBody = 'Hi @' + SERVER_JOBS_ADMIN + ', PTAL at this PR, ' +
+      'it adds/modifies a server job. Also @' + prAuthor + ', please ' +
+      'endeavour to fill the ' + serverJobsForm + ' for the new job ' +
```
Comment (ankita240796) [inline lib/checkPullRequestJob.js:45]: endeavour to fill the -> fill the.

Also, provide a link to the wiki page with instructions for new jobs: https://github.com/oppia/oppia/wiki/Running-jobs-in-production#submitting-a-pr-with-a-new-job.

File: lib/checkPullRequestJob.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+      'to get tested on the test server. Thanks!';
+    const commentParams = context.repo({
+      issue_number: pullRequest.number,
+      body: commentBody,
+    });
+
+    await context.github.issues.createComment(commentParams);
```
Comment (ankita240796) [inline lib/checkPullRequestJob.js:52]: We should also add the critical label to the PR.

File: lib/checkPullRequestJob.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
+const {SERVER_JOBS_ADMIN} = require('../userWhitelist.json');
+/**
+ * @param {string} filename
+ */
+const isJobFile = (filename) => {
```
Comment (ankita240796) [inline lib/checkPullRequestJob.js:5]: What if a job file is being modified but no new job is introduced? There can be a case when an issue is being fixed or some variables are being renamed. We only need the test procedure for new jobs.

Further, we need to have a check which ensures that whenever a new job is added, the name of the job is added to jobs_registry file.

File: lib/checkPullRequestJob.js (hunk 3 above)
Comment (ankita240796) [reply lib/checkPullRequestJob.js:45]: Also, can we mark such PRs as unmergeable until approved by Server jobs admin?

\cc @seanlip @nithusha21 @DubeySandeep

File: lib/checkPullRequestJob.js (hunk 1 above)
Comment (jameesjohn) [reply lib/checkPullRequestJob.js:7]: I wasn't sure of what to name since this isn't a prefix or suffix.

File: lib/checkPullRequestJob.js (hunk 5 above)
Comment (jameesjohn) [reply lib/checkPullRequestJob.js:5]: We may not be able to check the exact content of the job_registry file. What we can do is to ensure that a new job file is accompanied with a change in the job_registry.

File: lib/checkPullRequestJob.js (hunk 3 above)
Comment (jameesjohn) [reply lib/checkPullRequestJob.js:45]: I think what we should do here is to assign the pr to the server jobs admin.

File: lib/checkPullRequestJob.js (hunk 4 above)
Comment (jameesjohn) [reply lib/checkPullRequestJob.js]: Done

File: lib/checkPullRequestJob.js (hunk 2 above)
Comment (jameesjohn) [reply lib/checkPullRequestJob.js]: Done.

File: lib/checkPullRequestJob.js (hunk 1 above)
Comment (jameesjohn) [reply lib/checkPullRequestJob.js]: Done.

File: lib/checkPullRequestJob.js (hunk 5 above)
Comment (jameesjohn) [reply lib/checkPullRequestJob.js]: > We may not be able to check the exact content of the job_registry file. What we can do is to ensure that a new job file is accompanied with a change in the job_registry.

I've been able to do this.

**Comment by jameesjohn** [comment]
PTAL @ankita240796.
I've addressed all comments. The implementation here is significantly different from the initial one.

Thanks.

**Comment by jameesjohn** [comment]
I also got some inputs from @seanlip regarding the oppiabot message.

File: constants.js [hunk 6]
```diff
@@ -57,8 +49,8 @@ module.exports.wipCheck = wipCheck;
 module.exports.assigneeCheck = assigneeCheck;
 module.exports.mergeConflictCheck = mergeConflictCheck;
 module.exports.allMergeConflictCheck = allMergeConflictCheck;
+module.exports.jobCheck = jobCheck;
 
-module.exports.getChecksWhitelist = function() {
+module.exports.getChecksWhitelist = function () {
   return checksWhitelist;
 };
-
```
Comment (ankita240796) [inline constants.js:64]: Revert.

File: lib/checkPullRequestJob.js [hunk 7]
```diff
@@ -0,0 +1,210 @@
+const {SERVER_JOBS_ADMIN} = require('../userWhitelist.json');
```
Comment (ankita240796) [inline lib/checkPullRequestJob.js:1]: Space after brackets.

File: package.json [hunk 8]
```diff
@@ -44,7 +44,8 @@
   "nodemonConfig": {
     "exec": "npm start",
     "watch": [
-      ".env"
+      ".env",
+      "."
```
Comment (ankita240796) [inline package.json:48]: Why this change?

File: constants.js [hunk 9]
```diff
@@ -22,25 +23,16 @@ const checksWhitelist = {
     [labelEvent]: [],
     [synchronizeEvent]: [],
     [closeEvent]: [],
-    [editEvent]: []
+    [editEvent]: [],
```
Comment (ankita240796) [inline constants.js:26]: We should not have terminating `,` in an array list in JavaScript. Revert here and below.

File: lib/checkPullRequestJob.js [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+ * @param {string} filename
+ */
+const getJobNameFromFileName = (filename) => {
+  // Filename is the relative path to the file, we need to
+  // get the exact file name.
+  const substrings = filename.split('/');
+  const actualFileNameWithExtension = substrings[substrings.length - 1];
+  const extensionIndex = actualFileNameWithExtension.indexOf('.py');
+
+  return actualFileNameWithExtension.substring(0, extensionIndex);
```
Comment (ankita240796) [inline lib/checkPullRequestJob.js:29]: I am a bit confused on what this function is doing, are you obtaining the name of a job being added or the name of the file without any dir paths in it? Can you add some examples in comments and update the name of the function accordingly if it is being used to obtain file name without extensions.

File: lib/checkPullRequestJob.js [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+};
+
+/**
+ * @param {string} filename
+ * @param {import('probot').Octokit.PullsListFilesResponse} files
+ */
+const doesRegistryContainFile = (filename, files) => {
+  const jobName = getJobNameFromFileName(filename);
+
+  const registryFile = files.find(({filename}) => {
```
Comment (ankita240796) [inline lib/checkPullRequestJob.js:39]: `registryFile` -> `registryFileIsPresent`

File: lib/checkPullRequestJob.js [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+ * @param {import('probot').Context} context
+ * @returns {Promise<import('probot').Octokit.PullsListFilesResponseItem[]>}
+ */
+const getAllChangedFiles = async (context) => {
+  const changedFiles = [];
+  /**
+   * @type {import('probot').Octokit.PullsGetResponse}
+   */
+  const pullRequest = context.payload.pull_request;
+  const maxFilesPerPage = 100;
```
Comment (ankita240796) [inline lib/checkPullRequestJob.js:57]: Any specific reason to choose 100 here, just curious?

File: lib/checkPullRequestJob.js [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+const checkForNewJob = async (context) => {
+  /**
+   * @type {import('probot').Octokit.PullsGetResponse} pullRequest
+   */
+  const pullRequest = context.payload.pull_request;
+  if (!hasCriticalLabel(pullRequest)) {
+    const changedFiles = await getAllChangedFiles(context);
+
+    // Get new jobs that were created in the PR.
+    const newJobFiles = changedFiles.filter((file) => {
```
Comment (ankita240796) [inline lib/checkPullRequestJob.js:175]: We can have one more case, a new job being added in an existing job file, right?

File: lib/checkPullRequestJob.js [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+      newLineFeed + 'Thanks!');
+  }
+};
+
+/**
+ * @param {import('probot').Octokit.PullsGetResponse} pullRequest
+ */
+const hasCriticalLabel = (pullRequest) => {
+  const criticalLabel = pullRequest.labels.find(
+    (label) => label.name.toLowerCase() === 'critical'
```
Comment (ankita240796) [inline lib/checkPullRequestJob.js:157]: Declare `critical` as a constant and use here.

File: lib/checkPullRequestJob.js [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+  const registryReminder = getRegistryReminder(newJobFiles, changedFiles);
+  const jobNameString = getJobNameString(newJobFiles);
+  const newLineFeed = '<br>';
+  const serverJobsForm = 'server jobs form'.link(
+    'https://goo.gl/forms/XIj00RJ2h5L55XzU2');
+
+  // This function will never be called when there are no job files.
+  if (newJobFiles.length === 1) {
+    const serverAdminPing = 'Hi @' + SERVER_JOBS_ADMIN + ', PTAL at this PR, ' +
+      'it adds a new server job.' + jobNameString;
```
Comment (ankita240796) [inline lib/checkPullRequestJob.js:129]: `server job` -> `one off job`

File: lib/checkPullRequestJob.js [hunk 16]
```diff
@@ -0,0 +1,210 @@
...
+
+    const prAuthorPing = 'Also @' + prAuthor + registryReminder +
+    'please make sure to fill in the ' + serverJobsForm +
+    ' for the new job to be tested on the backup server. ' +
+    'This PR can be merged only after the test is successful.';
```
Comment (ankita240796) [inline lib/checkPullRequestJob.js:134]: Add a link to the guide on jobs:  https://github.com/oppia/oppia/wiki/Running-jobs-in-production#submitting-a-pr-with-a-new-job.

File: lib/checkPullRequestJob.js [hunk 17]
```diff
@@ -0,0 +1,210 @@
...
+    return (serverAdminPing + newLineFeed + prAuthorPing +
+      newLineFeed + 'Thanks!');
+  } else {
+    const serverAdminPing = 'Hi @' + SERVER_JOBS_ADMIN + ', PTAL at this PR, ' +
+      'it adds new server jobs.' + jobNameString;
+
+    const prAuthorPing = 'Also @' + prAuthor + registryReminder +
+    'please make sure to fill in the ' + serverJobsForm +
+    ' for the new jobs  to be tested on the backup server.' +
+    ' This PR can be merged only after the test is successful.';
```
Comment (ankita240796) [inline lib/checkPullRequestJob.js:145]: Ditto.

File: lib/checkPullRequestJob.js [hunk 18]
```diff
@@ -0,0 +1,210 @@
...
+  let registryReminder = ', ';
+
+  const hasNotAddedFileToRegistry = newJobFiles.some(
+    (file) => !doesRegistryContainFile(file.filename, changedFiles)
+  );
+
+  if (hasNotAddedFileToRegistry) {
+    const fileText = newJobFiles.length > 1 ? 'files' : 'file';
+    registryReminder +=
+      'endeavour to add the new job ' + fileText + ' to the job registry and ';
```
Comment (ankita240796) [inline lib/checkPullRequestJob.js:91]: Please add the new job...

File: spec/checkPullRequestJobSpec.js [hunk 19]
```diff
@@ -0,0 +1,401 @@
...
+    filename: 'core/domain/exp_jobs_oppiabot_off.py',
+    status: 'added',
+    additions: 1,
+    deletions: 0,
+    changes: 1,
+    blob_url:
+      'https://github.com/oppia/oppia/blob/67fb4a973b318882af3b5a894130e110d7e9833c/core/domain/exp_jobs_oppiabot_off.py',
+    raw_url:
+      'https://github.com/oppia/oppia/raw/67fb4a973b318882af3b5a894130e110d7e9833c/core/domain/exp_jobs_oppiabot_off.py',
+    contents_url:
```
Comment (ankita240796) [inline spec/checkPullRequestJobSpec.js:60]: Can we use this to obtain the content of the file?

File: lib/checkPullRequestJob.js (hunk 3 above)
Comment (ankita240796) [reply lib/checkPullRequestJob.js]: That makes sense. Let us do that.

File: lib/checkPullRequestJob.js (hunk 16 above)
Comment (jameesjohn) [reply lib/checkPullRequestJob.js]: Done.

File: lib/checkPullRequestJob.js (hunk 17 above)
Comment (jameesjohn) [reply lib/checkPullRequestJob.js]: Done.

File: spec/checkPullRequestJobSpec.js (hunk 19 above)
Comment (jameesjohn) [reply spec/checkPullRequestJobSpec.js:60]: Probably
The changes in the file are in the `patch`

File: lib/checkPullRequestJob.js (hunk 7 above)
Comment (ankita240796) [reply lib/checkPullRequestJob.js:1]: Missed this one.

File: constants.js (hunk 9 above)
Comment (ankita240796) [reply constants.js:26]: Ditto.

File: constants.js (hunk 6 above)
Comment (ankita240796) [reply constants.js:64]: Ditto.

File: lib/checkPullRequestJob.js (hunk 10 above)
Comment (ankita240796) [reply lib/checkPullRequestJob.js:30]: Can you add an example in the comments?

File: lib/checkPullRequestJob.js (hunk 11 above)
Comment (ankita240796) [reply lib/checkPullRequestJob.js:40]: Ok, Thanks for explaining.

File: lib/checkPullRequestJob.js (hunk 12 above)
Comment (ankita240796) [reply lib/checkPullRequestJob.js:58]: Add a comment.

File: lib/checkPullRequestJob.js (hunk 13 above)
Comment (ankita240796) [reply lib/checkPullRequestJob.js:179]: Yes.

File: package.json (hunk 8 above)
Comment (ankita240796) [reply package.json:48]: Ok, Thanks for doing this!

File: spec/checkPullRequestJobSpec.js (hunk 19 above)
Comment (ankita240796) [reply spec/checkPullRequestJobSpec.js:60]: Ok, so we can use `patch` for finding new jobs being added in an existing file, right?

File: spec/checkPullRequestJobSpec.js [hunk 20]
```diff
@@ -0,0 +1,410 @@
...
+const apiForSheetsModule = require('../lib/apiForSheets');
+const checkPullRequestLabelsModule = require('../lib/checkPullRequestLabels');
+const checkPullRequestBranchModule = require('../lib/checkPullRequestBranch');
+const checkWIPModule = require('../lib/checkWipDraftPR');
+const scheduler = require('../lib/scheduler');
+
+let payloadData = JSON.parse(
+  JSON.stringify(require('../fixtures/pullRequestPayload.json'))
+);
+const {SERVER_JOBS_ADMIN} = require('../userWhitelist.json');
```
Comment (ankita240796) [inline spec/checkPullRequestJobSpec.js:15]: Space after brackets.

File: spec/checkPullRequestJobSpec.js [hunk 21]
```diff
@@ -0,0 +1,410 @@
...
+    spyOn(apiForSheetsModule, 'checkClaStatus').and.callFake(() => {});
+    spyOn(
+      checkPullRequestLabelsModule,
+      'checkChangelogLabel'
+    ).and.callFake(() => {});
+    spyOn(checkPullRequestBranchModule, 'checkBranch').and.callFake(() => {});
+    spyOn(checkWIPModule, 'checkWIP').and.callFake(() => {});
+  });
+
+  describe('When pull request creates a new job file', () => {
```
Comment (ankita240796) [inline spec/checkPullRequestJobSpec.js:110]: `When pull request creates a new job file` -> `When a new job file is created in a pull request`

File: spec/checkPullRequestJobSpec.js [hunk 22]
```diff
@@ -0,0 +1,410 @@
...
+      expect(github.issues.addLabels).toHaveBeenCalledWith({
+        issue_number: payloadData.payload.pull_request.number,
+        repo: payloadData.payload.repository.name,
+        owner: payloadData.payload.repository.owner.login,
+        labels: ['critical']
+      });
+    });
+  });
+
+  describe('When pull request creates multiple job files', () => {
```
Comment (ankita240796) [inline spec/checkPullRequestJobSpec.js:178]: Ditto.

File: spec/checkPullRequestJobSpec.js [hunk 23]
```diff
@@ -0,0 +1,410 @@
...
+   * @type {import('probot').Octokit} github
+   */
+  let github;
+
+  /**
+   * @type {import('probot').Application} app
+   */
+  let app;
+
+  const newJobFileObj = {
```
Comment (ankita240796) [inline spec/checkPullRequestJobSpec.js:33]: firstNewJobFileObj

File: spec/checkPullRequestJobSpec.js [hunk 24]
```diff
@@ -0,0 +1,410 @@
...
+      expect(github.issues.addLabels).toHaveBeenCalledWith({
+        issue_number: payloadData.payload.pull_request.number,
+        repo: payloadData.payload.repository.name,
+        owner: payloadData.payload.repository.owner.login,
+        labels: ['critical']
+      });
+    });
+  });
+
+  describe('When pull request creates a new job file and updates registry', () => {
```
Comment (ankita240796) [inline spec/checkPullRequestJobSpec.js:248]: `When pull request creates a new job file and updates registry` -> `When  a new job file is created and registry is updated in a pull request`

File: spec/checkPullRequestJobSpec.js [hunk 25]
```diff
@@ -0,0 +1,410 @@
...
+      expect(github.issues.addLabels).toHaveBeenCalledWith({
+        issue_number: payloadData.payload.pull_request.number,
+        repo: payloadData.payload.repository.name,
+        owner: payloadData.payload.repository.owner.login,
+        labels: ['critical']
+      });
+    });
+  });
+
+  describe('When pull request modifies an existing job file', () => {
```
Comment (ankita240796) [inline spec/checkPullRequestJobSpec.js:317]: We will need to refactor this into two cases:

1. No new job being added in an existing job file
2. New job being added in an existing job file

File: spec/checkPullRequestJobSpec.js [hunk 26]
```diff
@@ -0,0 +1,410 @@
...
+    });
+    it('should not add critical label', () => {
+      expect(github.issues.addLabels).not.toHaveBeenCalled();
+    });
+    it('should not assign server jobs admin', () => {
+      expect(github.issues.addAssignees).not.toHaveBeenCalled();
+    });
+  });
+
+  describe('when pull request does not modify job file', () => {
```
Comment (ankita240796) [inline spec/checkPullRequestJobSpec.js:350]: When no job file is modified in a pull request

**Comment by ankita240796** [comment]
Please update the PR title and description to include Fixes #50.

File: constants.js (hunk 6 above)
Comment (ankita240796) [reply constants.js:64]: Ah ok I see, there was an extra line.

File: lib/checkPullRequestJob.js [hunk 27]
```diff
@@ -0,0 +1,266 @@
...
+
+  // A new job class should start with class and include OneOffJob in its name.
+  const newJobDefinitions = changesArray.filter(
+    (change) =>
+      change.startsWith(classIdentifier) && change.includes(oneOffJobIdentifier)
+  );
+
+  // New job definitions should look like:
+  // class OppiabotContributionsOneOffJob(jobs.BaseMapReduceOneOffJobManager):
+  // To get the actual job name, we need to remove the unneeded items.
```
Comment (ankita240796) [inline lib/checkPullRequestJob.js:40]: unneeded items -> class prefix and superclass

File: lib/checkPullRequestJob.js [hunk 28]
```diff
@@ -0,0 +1,266 @@
...
+  const extensionIndex = actualFileNameWithExtension.indexOf('.py');
+
+  return actualFileNameWithExtension.substring(0, extensionIndex);
+};
+
+/**
+ * @param {string} job
+ * @param {import('probot').Octokit.PullsListFilesResponse} files
+ */
+const doesRegistryContainFile = (job, files) => {
```
Comment (ankita240796) [inline lib/checkPullRequestJob.js:83]: Rename to `isRegistryUpdated` since it is checking two things whether registry is modified and if modified does it include the new job.

File: spec/checkPullRequestJobSpec.js (hunk 20 above)
Comment (ankita240796) [reply spec/checkPullRequestJobSpec.js:15]: I think you missed this.

File: lib/checkPullRequestJob.js (hunk 27 above)
Comment (jameesjohn) [reply lib/checkPullRequestJob.js]: Done.

File: lib/checkPullRequestJob.js (hunk 28 above)
Comment (jameesjohn) [reply lib/checkPullRequestJob.js]: Done.

File: spec/checkPullRequestJobSpec.js (hunk 20 above)
Comment (jameesjohn) [reply spec/checkPullRequestJobSpec.js]: Oops my bad.
Done.

**Comment by nishantwrp** [comment]
I've tested this functionality. Everything seems fine! But there is no comment when an existing job is modified as suggested in the description of this pull request.

**Comment by jameesjohn** [comment]
Thanks @nishantwrp.
> But there is no comment when an existing job is modified as suggested in the description of this pull request.

For existing job files, we are only commenting when the modifications includes a new job being added.

File: lib/checkPullRequestJob.js [hunk 29]
```diff
@@ -0,0 +1,266 @@
...
+  // '+' represents an addition while '-' represents a deletion.
+  const newLineAdded = '\n+';
+  const changesArray = file.patch.split(newLineAdded);
+  const classIdentifier = 'class';
+  const oneOffJobIdentifier = 'OneOffJob';
+
+  // A new job class should start with class and include OneOffJob in its name.
+  const newJobDefinitions = changesArray.filter(
+    (change) =>
+      change.startsWith(classIdentifier) && change.includes(oneOffJobIdentifier)
```
Comment (ankita240796) [inline lib/checkPullRequestJob.js:35]: Let us have a regex instead of this and grab the name using a match group. Would be more readable.

File: lib/checkPullRequestJob.js [hunk 30]
```diff
@@ -0,0 +1,266 @@
...
+ * @param {string} prAuthor - Author of the Pull Request
+ */
+const getCommentBody = (newJobFiles, changedFiles, prAuthor) => {
+  const registryReminder = getRegistryReminder(newJobFiles, changedFiles);
+  const jobNameString = getJobNameString(newJobFiles);
+  const newLineFeed = '<br>';
+  const serverJobsForm = 'server jobs form'.link(
+    'https://goo.gl/forms/XIj00RJ2h5L55XzU2');
+  const wikiLinkText = (
+    'This PR can be merged only after the test is successful'.link(
```
Comment (ankita240796) [inline lib/checkPullRequestJob.js:178]: test -> test run is successful. Please refer this guide [link to this guide] for details

File: lib/checkPullRequestJob.js [hunk 31]
```diff
@@ -0,0 +1,266 @@
...
+      const jobName = getJobNameFromFileName(file.filename);
+      return jobs.some((job) => !isRegistryUpdated(
+        `${jobName}.${job}`, changedFiles ));
+    }
+  );
+
+  if (hasNotAddedFileToRegistry) {
+    const fileText = newJobFiles.length > 1 ? 'files' : 'file';
+    registryReminder +=
+      'please add the new job ' + fileText + ' to the job registry and ';
```
Comment (ankita240796) [inline lib/checkPullRequestJob.js:143]: to the job registry [Link to jobs_registry.py]

File: lib/checkPullRequestJob.js (hunk 29 above)
Comment (jameesjohn) [reply lib/checkPullRequestJob.js]: Done.
Changing to regex helped reveal a bug so it was a good call.
Thanks.

File: lib/checkPullRequestJob.js (hunk 31 above)
Comment (jameesjohn) [reply lib/checkPullRequestJob.js]: Done.

File: lib/checkPullRequestJob.js (hunk 30 above)
Comment (jameesjohn) [reply lib/checkPullRequestJob.js]: Done.

File: lib/checkPullRequestJob.js [hunk 32]
```diff
@@ -0,0 +1,262 @@
...
+      const jobs = getNewJobsFromFile(file);
+      const jobName = getJobNameFromFileName(file.filename);
+      return jobs.some((job) => !isRegistryUpdated(
+        `${jobName}.${job}`, changedFiles ));
+    }
+  );
+
+  if (hasNotAddedFileToRegistry) {
+    const fileText = newJobFiles.length > 1 ? 'files' : 'file';
+    const jobRegistryLink = 'job registry'.link(
```
Comment (ankita240796) [inline lib/checkPullRequestJob.js:133]: job registry file.

File: lib/checkPullRequestJob.js [hunk 33]
```diff
@@ -0,0 +1,262 @@
...
+ * @param {string} prAuthor - Author of the Pull Request
+ */
+const getCommentBody = (newJobFiles, changedFiles, prAuthor) => {
+  const registryReminder = getRegistryReminder(newJobFiles, changedFiles);
+  const jobNameString = getJobNameString(newJobFiles);
+  const newLineFeed = '<br>';
+  const serverJobsForm = 'server jobs form'.link(
+    'https://goo.gl/forms/XIj00RJ2h5L55XzU2');
+  const wikiLinkText = (
+    'Please refer to this guide for details'.link(
```
Comment (ankita240796) [inline lib/checkPullRequestJob.js:171]: Can we just link to this guide instead of the complete line.

File: lib/checkPullRequestJob.js [hunk 34]
```diff
@@ -0,0 +1,262 @@
...
+
+  const newJobDefinitions = changesArray.filter(change => {
+    const matches = jobRegex.exec(change);
+    return matches !== null;
+  });
+
+  const newJobs = newJobDefinitions.map(definition => {
+    const matches = jobRegex.exec(definition);
+    return matches.groups.jobName + matches.groups.jobSuffix;
+  });
```
Comment (ankita240796) [inline lib/checkPullRequestJob.js:39]: Thanks for doing this!

Comment (jameesjohn) [reply lib/checkPullRequestJob.js:39]: Yeah...
Nitish's article helped a lot.

File: lib/checkPullRequestJob.js (hunk 32 above)
Comment (jameesjohn) [reply lib/checkPullRequestJob.js]: Done.

File: lib/checkPullRequestJob.js (hunk 33 above)
Comment (jameesjohn) [reply lib/checkPullRequestJob.js]: Done.

**Comment by jameesjohn** [comment]
Hi @ankita240796, this is ready to be merged right?

File: lib/checkPullRequestJob.js (hunk 34 above)
Comment (ankita240796) [reply lib/checkPullRequestJob.js:39]: @bansalnitish ping :D

**Comment by ankita240796** [comment]
Yes, thanks @jameesjohn! Thanks for doing this so well!
